### DISCUSSION TITLE
Python syntax question from a Java developer...

### DISCUSSION THREAD

**Discussion by kazetsukaimiko** [jblance/mpp-solar#127] [General]
I am a Java developer looking at the protocol documentation for the DALY BMS for a similar, multi-device integration project:
https://github.com/kazetsukaimiko/freedriver (where the daly library will live)
https://github.com/kazetsukaimiko/autonomy (the J2EE microservices that will use it)

I'm using your implementation as reference, and had a general python question. You have some test_responses in your daly.py protocol definitions:

```python
            "test_responses": [
            b"\xa5\x01\x90\x08\x02\x10\x00\x00uo\x03\xbc\xf3",
            b"\xa5\x01\x90\x08\x02\x14\x00\x00uE\x03x\x89",
            b"\xa5\x01\x90\x08\x01\t\x00\x00u\xcf\x03\n\x99",
        ],
```

I kinda sorta see hexadecimal/binary encoding here, but I the parts I'm having trouble understanding are:
\x00uE
\x00u
\x00uo
\n
\t

I'm trying to make an equivalent in Java to build tests off:
```java
    int[] SOC_1 = new int[]{
            0xA5, 0x01, 0x90, 0x08, 0x02, 0x10, 0x00, 0x00, 0x03, 0xBC, 0xF3
    };
```

Would you mind helping me understand what's going on there? Your code is a great reference, and the first I've seen with examples for responses.

**Comment by jblance** [comment]
Hi

They are just characters that have an ascii representation of the hex number.
i.e `\x00uE` is `\x00\x75\x45`

```python
> print(b'\x00\x75\x45')
b'\x00uE'
 ```

likewise `\n` is the same as `\x0a`

**Reply by kazetsukaimiko** [reply]
Thank you for this. Ultimately I used something to the effect of:
`for byte in  b"\xa5\x01\x90\x08\x02\x10\x00\x00uo\x03\xbc\xf3": print(hex(byte))`

To get the same byte array in Java.

I will say peppering in ASCII makes it a little harder to read in this regard.

Using your python crc8 function, I was able to write a unit test in Java that made sure my algorithm matched yours. It is rather simple but I'm not very used to doing binary-level operations. There were lots of "CRC8" functions out there in Java that did not produce identical results.

Finally, would you mind either pointing me at some code that shows examples of how to make requests to the BMS? Daly's PDF documentation doesn't make it too clear to me what the request payload should look like. I know we start with:
0xA5 (start flag)
0x80 (coming from the "Upper" / computer)
[Data Id]

What is unclear to me is what/if we send for Data Length and Data fields. For getting information I would guess all-zeroes followed by the checksum, for setting BMS settings I would guess data length and payload would be used in some way to make changes (although then the password would have to be introduced somewhere).

Really appreciate the help.

**Reply by jblance** [reply]
its not really 'peppering in ascii' - its a byte array where some of the bytes can only be represented in hex. They are also copied from actual debugs, so I dont really need them to be readible the are just for testing

there are some DALY debugs in the issues, eg #113 

one example `full command b'\xa5@\x95\x08\x00\x00\x00\x00\x00\x00\x00\x00\x82\n' for command cellVoltages`
